Userland/Libraries/LibJS/Runtime/Value.h

```diff
@@ -343,35 +438,78 @@ class Value {
     [[nodiscard]] ALWAYS_INLINE ThrowCompletionOr<Value> invoke(GlobalObject& global_object, PropertyKey const& property_key, Args... args);
 
 private:
-    Type m_type { Type::Empty };
+    //    Type type() const { return Type::Empty; }
+
+    Value(u64 tag, u64 val)
+    {
+        //        dbgln("creating value with {:X} | {:X}", tag, val);
+        VERIFY(!(tag & val));
+        m_value.encoded = tag | val;
+    }
+
+    Value(u64 tag, void const* ptr)
+    {
+        //        dbgln("creating ptr value with {:X} | {:X}", tag, ptr);
+        VERIFY((tag & 0x8000000000000000ul) == 0 || ptr);
+        VERIFY(!(reinterpret_cast<u64>(ptr) & (0xFFFFul) << 48));
+        VERIFY(!(reinterpret_cast<u64>(ptr) & 0x3));
+        m_value.encoded = tag | reinterpret_cast<u64>(ptr);
+    }
+
+    bool is_double() const { return (m_value.encoded & DOUBLE_EXTRACT_TAG) != DOUBLE_EXTRACT_TAG || (m_value.encoded & 0x7FFFFFFFFFFFFFFFul) == CANON_NAN_BITS; }
+    bool is_int32() const { return (m_value.encoded & TAG_EXTRACT_TAG) == INT32_TAG; }
+
+    i32 as_i32() const
+    {
+        VERIFY(is_int32());
+        return static_cast<i32>(m_value.encoded & 0xFFFFFFFF);
+    }
+
+    template<typename PointerType>
+    PointerType const* extract_pointer() const
+    {
+        VERIFY(is_cell());
+        u64 ptr_val = m_value.encoded & ((1ull << 48) - 1);
+        VERIFY(ptr_val);
+        return reinterpret_cast<PointerType*>(ptr_val);
+    }
+
+    template<typename PointerType>
+    PointerType* extract_pointer()
+    {
+        VERIFY(is_cell());
+        u64 ptr_val = m_value.encoded & ((1ull << 48) - 1);
+        VERIFY(ptr_val);
+        return reinterpret_cast<PointerType*>(ptr_val);
+    }
 
     [[nodiscard]] ThrowCompletionOr<Value> invoke_internal(GlobalObject& global_object, PropertyKey const&, Optional<MarkedVector<Value>> arguments);
 
     ThrowCompletionOr<i32> to_i32_slow_case(GlobalObject&) const;
 
     union {
-        bool as_bool;
-        i32 as_i32;
         double as_double;
-        PrimitiveString* as_string;
-        Symbol* as_symbol;
-        Object* as_object;
-        Cell* as_cell;
-        Accessor* as_accessor;
-        BigInt* as_bigint;
-
         u64 encoded;
```

Comment:
Making this a bit field would avoid some of the magic bit shifts and ands
Also then putting the double diagram and consorts closer to this would make it easier to understand the future
Example:
```suggestion
        struct {
            u64 sign : 1;
            u64 exponent :  11;
            u64 tag : 5;
            u64 pointer : 47;
        };
        u64 as_raw_u64;
```
adjust to the order actual values you are using
(also idk if the order needs to be reversed, bit fields are weird)

Reply:
Thanks, it does make the code more readable (and yes its the opposite order)